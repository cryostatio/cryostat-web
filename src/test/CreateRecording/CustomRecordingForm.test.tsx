/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import * as React from 'react';
import { createMemoryHistory } from 'history';
import { screen, cleanup, act as doAct } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';
import { ServiceContext, Services } from '@app/Shared/Services/Services';
import { defaultServices } from '@app/Shared/Services/Services';
import { EventTemplate, RecordingAttributes, RecordingOptions } from '@app/Shared/Services/Api.service';
import { of, Subject } from 'rxjs';
import { renderWithServiceContext } from '../Common';

jest.mock('@patternfly/react-core', () => ({
  // Mock out tooltip for snapshot testing
  ...jest.requireActual('@patternfly/react-core'),
  Tooltip: ({ children }) => <>{children}</>,
}));

import { CustomRecordingForm } from '@app/CreateRecording/CustomRecordingForm';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';
import { TargetService } from '@app/Shared/Services/Target.service';

const mockConnectUrl = 'service:jmx:rmi://someUrl';
const mockTarget = { connectUrl: mockConnectUrl, alias: 'fooTarget' };

const mockCustomEventTemplate: EventTemplate = {
  name: 'someEventTemplate',
  description: 'Some Description',
  provider: 'Cryostat',
  type: 'CUSTOM',
};

const mockRecordingOptions: RecordingOptions = {
  toDisk: true,
  maxAge: undefined,
  maxSize: 0,
};

const mockResponse: Response = {
  ok: true,
} as Response;

jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));
jest
  .spyOn(defaultServices.api, 'doGet')
  .mockReturnValueOnce(of([mockCustomEventTemplate])) // renders correctly
  .mockReturnValueOnce(of(mockRecordingOptions))

  .mockReturnValueOnce(of([mockCustomEventTemplate])) // should create recording when form is filled and create is clicked
  .mockReturnValueOnce(of(mockRecordingOptions))

  .mockReturnValueOnce(of([mockCustomEventTemplate])) // should show correct helper texts in metadata label editor
  .mockReturnValueOnce(of(mockRecordingOptions));

jest.spyOn(defaultServices.target, 'authFailure').mockReturnValue(of());

const history = createMemoryHistory({ initialEntries: ['/recordings/create'] });

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: () => ({ url: history.location.pathname }),
  useHistory: () => history,
}));

describe('<CustomRecordingForm />', () => {
  beforeEach(() => {
    history.go(-history.length);
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <CustomRecordingForm />
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should create recording when form is filled and create is clicked', async () => {
    const onSubmitSpy = jest.spyOn(defaultServices.api, 'createRecording').mockReturnValue(of(mockResponse));
    const { user } = renderWithServiceContext(<CustomRecordingForm />);

    const nameInput = screen.getByLabelText('Name *');
    expect(nameInput).toBeInTheDocument();
    expect(nameInput).toBeVisible();

    const templateSelect = screen.getByLabelText('Template *');
    expect(templateSelect).toBeInTheDocument();
    expect(templateSelect).toBeVisible();

    await user.type(nameInput, 'a_recording');
    await user.selectOptions(templateSelect, ['someEventTemplate']);

    const createButton = screen.getByRole('button', { name: /^create$/i });
    expect(createButton).toBeInTheDocument();
    expect(createButton).toBeVisible();

    expect(createButton).not.toBeDisabled();
    await user.click(createButton);

    expect(onSubmitSpy).toHaveBeenCalledTimes(1);
    expect(onSubmitSpy).toHaveBeenCalledWith({
      name: 'a_recording',
      events: 'template=someEventTemplate,type=CUSTOM',
      duration: 30,
      archiveOnStop: true,
      options: {
        toDisk: true,
        maxAge: undefined,
        maxSize: 0,
      },
      metadata: { labels: {} },
    } as RecordingAttributes);
  });

  it('should show correct helper texts in metadata label editor', async () => {
    const { user } = renderWithServiceContext(<CustomRecordingForm />);

    const metadataEditorToggle = screen.getByText('Show metadata options');
    expect(metadataEditorToggle).toBeInTheDocument();
    expect(metadataEditorToggle).toBeVisible();

    await user.click(metadataEditorToggle);

    const helperText = screen.getByText(/are set by Cryostat and will be overwritten if specifed\.$/);
    expect(helperText).toBeInTheDocument();
    expect(helperText).toBeVisible();
  });

  it('should show error view if failing to retrieve templates or recording options', async () => {
    const subj = new Subject<void>();
    const mockTargetSvc = {
      target: () => of(mockTarget),
      authFailure: () => subj.asObservable(),
    } as TargetService;
    const services: Services = {
      ...defaultServices,
      target: mockTargetSvc,
    };
    renderWithServiceContext(<CustomRecordingForm />, { services: services });

    await doAct(async () => subj.next());

    const failTitle = screen.getByText('Error displaying recording creation form');
    expect(failTitle).toBeInTheDocument();
    expect(failTitle).toBeVisible();

    const authFailText = screen.getByText('Auth failure');
    expect(authFailText).toBeInTheDocument();
    expect(authFailText).toBeVisible();

    const retryButton = screen.getByText('Retry');
    expect(retryButton).toBeInTheDocument();
    expect(retryButton).toBeVisible();
  });
});
