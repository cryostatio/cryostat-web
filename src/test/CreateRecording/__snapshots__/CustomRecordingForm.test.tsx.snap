// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomRecordingForm /> renders correctly 1`] = `
Array [
  <small
    className=""
    data-ouia-component-id="OUIA-Generated-Text-1"
    data-ouia-component-type="PF5/Text"
    data-ouia-safe={true}
    data-pf-content={true}
  >
    JDK Flight Recordings are compact records of events which have occurred within the Target JVM. Many event types are built in to the JVM itself, while others are user defined.
  </small>,
  <form
    className="pf-v5-c-form pf-m-horizontal"
    noValidate={true}
  >
    <div
      className="pf-v5-c-form__group"
    >
      <div
        className="pf-v5-c-form__group-label"
      >
        <label
          className="pf-v5-c-form__label"
          htmlFor="recording-name"
        >
          <span
            className="pf-v5-c-form__label-text"
          >
            Name
          </span>
          <span
            aria-hidden="true"
            className="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        className="pf-v5-c-form__group-control"
      >
        <span
          className="pf-v5-c-form-control"
        >
          <input
            aria-describedby="recording-name-helper"
            aria-invalid={false}
            aria-label={null}
            data-ouia-component-id="OUIA-Generated-TextInputBase-1"
            data-ouia-component-type="PF5/TextInput"
            data-ouia-safe={true}
            data-quickstart-id="crf-name"
            disabled={false}
            id="recording-name"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
        </span>
        <div
          className="pf-v5-c-check"
        >
          <input
            aria-invalid={false}
            aria-label="restartExisting checkbox"
            checked={false}
            className="pf-v5-c-check__input"
            data-ouia-component-id="OUIA-Generated-Checkbox-1"
            data-ouia-component-type="PF5/Checkbox"
            data-ouia-safe={true}
            disabled={false}
            id="recording-restart-existing"
            name="recording-restart-existing"
            onChange={[Function]}
            required={false}
            type="checkbox"
          />
          <label
            className="pf-v5-c-check__label"
            htmlFor="recording-restart-existing"
          >
            Restart if Recording already exists
          </label>
        </div>
        <div
          className="pf-v5-c-form__helper-text"
        >
          <div
            className="pf-v5-c-helper-text"
          >
            <div
              className="pf-v5-c-helper-text__item"
            >
              <span
                className="pf-v5-c-helper-text__item-text"
              >
                Enter a Recording name. This will be unique within the target JVM.
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-form__group"
      data-quickstart-id="crf-duration"
    >
      <div
        className="pf-v5-c-form__group-label"
      >
        <label
          className="pf-v5-c-form__label"
          htmlFor="recording-duration"
        >
          <span
            className="pf-v5-c-form__label-text"
          >
            Duration
          </span>
          <span
            aria-hidden="true"
            className="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        className="pf-v5-c-form__group-control"
      >
        <div
          className="pf-v5-l-split pf-m-gutter"
        >
          <div
            className="pf-v5-l-split__item"
          >
            <div
              className="pf-v5-c-check"
            >
              <input
                aria-invalid={false}
                aria-label="Continuous checkbox"
                checked={false}
                className="pf-v5-c-check__input"
                data-ouia-component-id="OUIA-Generated-Checkbox-2"
                data-ouia-component-type="PF5/Checkbox"
                data-ouia-safe={true}
                disabled={false}
                id="recording-continuous"
                name="recording-continuous"
                onChange={[Function]}
                required={false}
                type="checkbox"
              />
              <label
                className="pf-v5-c-check__label"
                htmlFor="recording-continuous"
              >
                Continuous
              </label>
            </div>
          </div>
          <div
            className="pf-v5-l-split__item"
          >
            <div
              className="pf-v5-c-check"
            >
              <input
                aria-invalid={false}
                aria-label="ArchiveOnStop checkbox"
                checked={true}
                className="pf-v5-c-check__input"
                data-ouia-component-id="OUIA-Generated-Checkbox-3"
                data-ouia-component-type="PF5/Checkbox"
                data-ouia-safe={true}
                disabled={false}
                id="recording-archive-on-stop"
                name="recording-archive-on-stop"
                onChange={[Function]}
                required={false}
                type="checkbox"
              />
              <label
                className="pf-v5-c-check__label"
                htmlFor="recording-archive-on-stop"
              >
                Archive on Stop
              </label>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-l-split pf-m-gutter"
        >
          <div
            className="pf-v5-l-split__item pf-m-fill"
          >
            <span
              className="pf-v5-c-form-control"
            >
              <input
                aria-invalid={false}
                aria-label="Duration Picker Period Input"
                data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                data-ouia-component-type="PF5/TextInput"
                data-ouia-safe={true}
                disabled={false}
                id="duration-picker-period"
                min="0"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="number"
                value={30}
              />
            </span>
          </div>
          <div
            className="pf-v5-l-split__item"
          >
            <span
              className="pf-v5-c-form-control"
            >
              <select
                aria-invalid={false}
                aria-label="Duration Picker Units Input"
                data-ouia-component-id="OUIA-Generated-FormSelect-default-1"
                data-ouia-component-type="PF5/FormSelect"
                data-ouia-safe={true}
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                value={1000}
              >
                <option
                  className=""
                  disabled={false}
                  value={1000}
                >
                  Seconds
                </option>
                <option
                  className=""
                  disabled={false}
                  value={60000}
                >
                  Minutes
                </option>
                <option
                  className=""
                  disabled={false}
                  value={3600000}
                >
                  Hours
                </option>
              </select>
              <span
                className="pf-v5-c-form-control__utilities"
              >
                <span
                  className="pf-v5-c-form-control__toggle-icon"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </div>
        </div>
        <div
          className="pf-v5-c-form__helper-text"
        >
          <div
            className="pf-v5-c-helper-text"
          >
            <div
              className="pf-v5-c-helper-text__item pf-m-success"
            >
              <span
                className="pf-v5-c-helper-text__item-text"
              >
                Time before the Recording is automatically stopped and copied to archive.
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-form__group"
    >
      <div
        className="pf-v5-c-form__group-label"
      >
        <label
          className="pf-v5-c-form__label"
          htmlFor="recording-template"
        >
          <span
            className="pf-v5-c-form__label-text"
          >
            Template
          </span>
          <span
            aria-hidden="true"
            className="pf-v5-c-form__label-required"
          >
             
            *
          </span>
        </label>
         
      </div>
      <div
        className="pf-v5-c-form__group-control"
      >
        <span
          className="pf-v5-c-form-control pf-m-placeholder"
        >
          <select
            aria-invalid={false}
            aria-label="Event Template Input"
            data-ouia-component-id="OUIA-Generated-FormSelect-default-2"
            data-ouia-component-type="PF5/FormSelect"
            data-ouia-safe={true}
            data-quickstart-id="template-selector"
            disabled={false}
            id="recording-template"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            value=""
          >
            <option
              className=""
              disabled={false}
              value=""
            >
              Select a Template
            </option>
            <optgroup
              className=""
              disabled={false}
              label="Target templates"
            />
            <optgroup
              className=""
              disabled={false}
              label="Custom Templates"
            >
              <option
                className=""
                disabled={false}
                value="someEventTemplate,CUSTOM"
              >
                someEventTemplate
              </option>
            </optgroup>
          </select>
          <span
            className="pf-v5-c-form-control__utilities"
          >
            <span
              className="pf-v5-c-form-control__toggle-icon"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                className="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </span>
        </span>
        <div
          className="pf-v5-c-form__helper-text"
        >
          <div
            className="pf-v5-c-helper-text"
          >
            <div
              className="pf-v5-c-helper-text__item"
            >
              <span
                className="pf-v5-c-helper-text__item-text"
              >
                A Template must be selected
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-expandable-section"
      data-quickstart-id="crf-metadata-opt"
    >
      <button
        aria-controls="expandable-section-content-1717748449488ubkqy21mox"
        className="pf-v5-c-expandable-section__toggle"
        id="expandable-section-toggle-1717748449488ye8g4ku6ktg"
        onClick={[Function]}
        type="button"
      >
        <span
          className="pf-v5-c-expandable-section__toggle-icon"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            className="pf-v5-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 256 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </span>
        <span
          className="pf-v5-c-expandable-section__toggle-text"
        >
          Show metadata options
        </span>
      </button>
      <div
        aria-labelledby="expandable-section-toggle-1717748449488ye8g4ku6ktg"
        className="pf-v5-c-expandable-section__content"
        hidden={true}
        id="expandable-section-content-1717748449488ubkqy21mox"
        role="region"
      >
        <div
          className="pf-v5-c-form__group"
        >
          <div
            className="pf-v5-c-form__group-label"
          >
            <label
              className="pf-v5-c-form__label"
              htmlFor="labels"
            >
              <span
                className="pf-v5-c-form__label-text"
              >
                Labels
              </span>
            </label>
             
            <span
              className="pf-v5-c-icon"
            >
              <span
                className="pf-v5-c-icon__content"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  className="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                  />
                </svg>
              </span>
            </span>
          </div>
          <div
            className="pf-v5-c-form__group-control"
          >
            <div
              className="pf-v5-c-form__helper-text"
            >
              <div
                className="pf-v5-c-helper-text"
              >
                <div
                  className="pf-v5-c-helper-text__item pf-m-dynamic"
                >
                  <span
                    className="pf-v5-c-helper-text__item-text"
                  >
                    Labels with key 
                    <span
                      className="pf-v5-c-label pf-m-compact"
                    >
                      <span
                        className="pf-v5-c-label__content"
                      >
                        <span
                          className="pf-v5-c-label__text"
                        >
                          template.name
                        </span>
                      </span>
                    </span>
                     and 
                    <span
                      className="pf-v5-c-label pf-m-compact"
                    >
                      <span
                        className="pf-v5-c-label__content"
                      >
                        <span
                          className="pf-v5-c-label__text"
                        >
                          template.type
                        </span>
                      </span>
                    </span>
                     are set by Cryostat and will be overwritten if specifed.
                    <span
                      className="pf-v5-screen-reader"
                    >
                      : 
                      default status
                      ;
                    </span>
                  </span>
                </div>
              </div>
            </div>
            <button
              aria-disabled={false}
              aria-label="Add Label"
              className="pf-v5-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <span
                className="pf-v5-c-button__icon pf-m-start"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  className="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                  />
                </svg>
              </span>
              Add Label
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-expandable-section"
      data-quickstart-id="crf-advanced-opt"
    >
      <button
        aria-controls="expandable-section-content-1717748449489nsjfqirxt"
        className="pf-v5-c-expandable-section__toggle"
        id="expandable-section-toggle-17177484494894h5omcs1uio"
        onClick={[Function]}
        type="button"
      >
        <span
          className="pf-v5-c-expandable-section__toggle-icon"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            className="pf-v5-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 256 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </span>
        <span
          className="pf-v5-c-expandable-section__toggle-text"
        >
          Show advanced options
        </span>
      </button>
      <div
        aria-labelledby="expandable-section-toggle-17177484494894h5omcs1uio"
        className="pf-v5-c-expandable-section__content"
        hidden={true}
        id="expandable-section-content-1717748449489nsjfqirxt"
        role="region"
      >
        <small
          className=""
          data-ouia-component-id="OUIA-Generated-Text-2"
          data-ouia-component-type="PF5/Text"
          data-ouia-safe={true}
          data-pf-content={true}
        >
          A value of 0 for maximum size or age means unbounded.
        </small>
        <div
          className="pf-v5-c-form__group"
        >
          <div
            className="pf-v5-c-form__group-control"
          >
            <div
              className="pf-v5-c-check"
            >
              <input
                aria-invalid={false}
                checked={true}
                className="pf-v5-c-check__input"
                data-ouia-component-id="OUIA-Generated-Checkbox-4"
                data-ouia-component-type="PF5/Checkbox"
                data-ouia-safe={true}
                disabled={false}
                id="toDisk-checkbox"
                onChange={[Function]}
                required={false}
                type="checkbox"
              />
              <label
                className="pf-v5-c-check__label"
                htmlFor="toDisk-checkbox"
              >
                To disk
              </label>
            </div>
            <div
              className="pf-v5-c-form__helper-text"
            >
              <div
                className="pf-v5-c-helper-text"
              >
                <div
                  className="pf-v5-c-helper-text__item"
                >
                  <span
                    className="pf-v5-c-helper-text__item-text"
                  >
                    Write contents of buffer onto disk. If disabled, the buffer acts as circular buffer only keeping the most recent Recording information
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-form__group"
        >
          <div
            className="pf-v5-c-form__group-label"
          >
            <label
              className="pf-v5-c-form__label"
              htmlFor="maxSize"
            >
              <span
                className="pf-v5-c-form__label-text"
              >
                Maximum size
              </span>
            </label>
             
          </div>
          <div
            className="pf-v5-c-form__group-control"
          >
            <div
              className="pf-v5-l-split pf-m-gutter"
            >
              <div
                className="pf-v5-l-split__item pf-m-fill"
              >
                <span
                  className="pf-v5-c-form-control"
                >
                  <input
                    aria-invalid={false}
                    aria-label="max size value"
                    data-ouia-component-id="OUIA-Generated-TextInputBase-3"
                    data-ouia-component-type="PF5/TextInput"
                    data-ouia-safe={true}
                    disabled={false}
                    id="maxSize"
                    min="0"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="number"
                    value={0}
                  />
                </span>
              </div>
              <div
                className="pf-v5-l-split__item"
              >
                <span
                  className="pf-v5-c-form-control"
                >
                  <select
                    aria-invalid={false}
                    aria-label="Max size units input"
                    data-ouia-component-id="OUIA-Generated-FormSelect-default-3"
                    data-ouia-component-type="PF5/FormSelect"
                    data-ouia-safe={true}
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    value={1}
                  >
                    <option
                      className=""
                      disabled={false}
                      value="1"
                    >
                      B
                    </option>
                    <option
                      className=""
                      disabled={false}
                      value={1024}
                    >
                      KiB
                    </option>
                    <option
                      className=""
                      disabled={false}
                      value={1048576}
                    >
                      MiB
                    </option>
                  </select>
                  <span
                    className="pf-v5-c-form-control__utilities"
                  >
                    <span
                      className="pf-v5-c-form-control__toggle-icon"
                    >
                      <svg
                        aria-hidden={true}
                        aria-labelledby={null}
                        className="pf-v5-svg"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        viewBox="0 0 320 512"
                        width="1em"
                      >
                        <path
                          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
              </div>
            </div>
            <div
              className="pf-v5-c-form__helper-text"
            >
              <div
                className="pf-v5-c-helper-text"
              >
                <div
                  className="pf-v5-c-helper-text__item"
                >
                  <span
                    className="pf-v5-c-helper-text__item-text"
                  >
                    The maximum size of Recording data saved to disk
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-form__group"
        >
          <div
            className="pf-v5-c-form__group-label"
          >
            <label
              className="pf-v5-c-form__label"
              htmlFor="maxAge"
            >
              <span
                className="pf-v5-c-form__label-text"
              >
                Maximum age
              </span>
            </label>
             
          </div>
          <div
            className="pf-v5-c-form__group-control"
          >
            <div
              className="pf-v5-l-split pf-m-gutter"
            >
              <div
                className="pf-v5-l-split__item pf-m-fill"
              >
                <span
                  className="pf-v5-c-form-control pf-m-disabled"
                >
                  <input
                    aria-invalid={false}
                    aria-label="Max age duration"
                    data-ouia-component-id="OUIA-Generated-TextInputBase-4"
                    data-ouia-component-type="PF5/TextInput"
                    data-ouia-safe={true}
                    disabled={true}
                    id="maxAgeDuration"
                    min="0"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="number"
                    value={0}
                  />
                </span>
              </div>
              <div
                className="pf-v5-l-split__item"
              >
                <span
                  className="pf-v5-c-form-control pf-m-disabled"
                >
                  <select
                    aria-invalid={false}
                    aria-label="Max Age units Input"
                    data-ouia-component-id="OUIA-Generated-FormSelect-default-4"
                    data-ouia-component-type="PF5/FormSelect"
                    data-ouia-safe={true}
                    disabled={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    value={1}
                  >
                    <option
                      className=""
                      disabled={false}
                      value="1"
                    >
                      Seconds
                    </option>
                    <option
                      className=""
                      disabled={false}
                      value={60}
                    >
                      Minutes
                    </option>
                    <option
                      className=""
                      disabled={false}
                      value={3600}
                    >
                      Hours
                    </option>
                  </select>
                  <span
                    className="pf-v5-c-form-control__utilities"
                  >
                    <span
                      className="pf-v5-c-form-control__toggle-icon"
                    >
                      <svg
                        aria-hidden={true}
                        aria-labelledby={null}
                        className="pf-v5-svg"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        viewBox="0 0 320 512"
                        width="1em"
                      >
                        <path
                          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
              </div>
            </div>
            <div
              className="pf-v5-c-form__helper-text"
            >
              <div
                className="pf-v5-c-helper-text"
              >
                <div
                  className="pf-v5-c-helper-text__item"
                >
                  <span
                    className="pf-v5-c-helper-text__item-text"
                  >
                    The maximum age of Recording data stored to disk
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-c-form__group pf-m-action"
    >
      <div
        className="pf-v5-c-form__group-control"
      >
        <div
          className="pf-v5-c-form__actions"
        >
          <button
            aria-disabled={true}
            aria-label={null}
            className="pf-v5-c-button pf-m-primary pf-m-disabled pf-m-progress"
            data-ouia-component-id="OUIA-Generated-Button-primary-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe={true}
            data-quickstart-id="crf-create-btn"
            disabled={true}
            onClick={[Function]}
            role={null}
            tabIndex={null}
            type="button"
          >
            Create
          </button>
          <button
            aria-disabled={false}
            aria-label={null}
            className="pf-v5-c-button pf-m-secondary"
            data-ouia-component-id="OUIA-Generated-Button-secondary-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe={true}
            disabled={false}
            onClick={[Function]}
            role={null}
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </form>,
]
`;
